evm_arithmetization/src/cpu/kernel/asm/memory/syscalls.asm

```diff
@@ -225,6 +225,12 @@ global sys_mcopy:
     // stack: kexit_info, dest_offset, offset, size
     %wcopy_charge_gas
 
+    // stack: kexit_info, dest_offset, offset, size
+    DUP2
+    %ensure_offset_in_range
+    DUP3
+    %ensure_offset_in_range
```

Comment:
Not directly related to this PR, but I noticed we call `%update_mem_bytes` only on `dest_offset+size` below.
My understanding of the EIP, especially the line 
```
If length > 0 and (src + length or dst + length) is beyond the current memory length, the memory is extended with respective gas cost applied.
```
would imply we also need to call it with `offset+size`.

Reply:
Hmm actually I think this would also be handling the error in that particular txn case. Will update the PR to have it.

Reply:
I also think there may be issues with the current `mcopy_with_overlap`. I don't think the current impl is valid if `SRC < DST` and `DST - SRC < min(32, size)`. To alleviate it without too much overhead, I guess we could simply have a `%memcpy_bytes_reversed` macro that processes the copy backwards, starting from the last offset to read / write from and decrementing both `DST`, `SRC` and `size`.